sentry/src/main/java/io/sentry/Breadcrumb.java

```diff
@@ -188,6 +188,113 @@ public static Breadcrumb fromMap(
     return breadcrumb;
   }
 
+  /**
+   * Creates a breadcrumb for a GraphQL operation.
+   *
+   * @param operationName - the name of the GraphQL operation
+   * @param operationType - the type of GraphQL operation (e.g. query, mutation, subscription)
+   * @param operationId - the ID of the GraphQL operation
+   * @return the breadcrumb
+   */
+  public static @NotNull Breadcrumb graphqlOperation(
+      final @Nullable String operationName,
+      final @Nullable String operationType,
+      final @Nullable String operationId) {
+    final Breadcrumb breadcrumb = new Breadcrumb();
+
+    breadcrumb.setType("graphql");
+
+    if (operationName != null) {
+      breadcrumb.setData("operation_name", operationName);
+    }
+    if (operationType != null) {
+      breadcrumb.setData("operation_type", operationType);
+      breadcrumb.setCategory(operationType);
+    } else {
+      breadcrumb.setCategory("graphql.operation");
+    }
+    if (operationId != null) {
+      breadcrumb.setData("operation_id", operationId);
+    }
+
+    return breadcrumb;
+  }
+
+  /**
+   * Creates a breadcrumb for a GraphQL data fetcher.
+   *
+   * @param path - the name of the GraphQL operation
+   * @param field - the type of GraphQL operation (e.g. query, mutation, subscription)
+   * @param type - the ID of the GraphQL operation
+   * @param objectType - the object type of the GraphQL data fetch operation
+   * @return the breadcrumb
+   */
+  public static @NotNull Breadcrumb graphqlDataFetcher(
+      final @Nullable String path,
+      final @Nullable String field,
+      final @Nullable String type,
+      final @Nullable String objectType) {
+    final Breadcrumb breadcrumb = new Breadcrumb();
+
+    breadcrumb.setType("graphql");
+    breadcrumb.setCategory("data_fetcher");
+
+    if (path != null) {
+      // TODO key?
+      breadcrumb.setData("graphql.path", path);
+    }
+    if (field != null) {
+      // TODO key?
+      breadcrumb.setData("graphql.field", field);
+    }
+    if (type != null) {
+      // TODO key?
+      breadcrumb.setData("graphql.type", type);
+    }
+    if (objectType != null) {
+      // TODO key?
+      breadcrumb.setData("graphql.object_type", objectType);
+    }
+
+    return breadcrumb;
+  }
+
+  /**
+   * Creates a breadcrumb for a GraphQL data loader.
+   *
+   * @param keys - keys to be fetched by the data loader
+   * @param keyType - class of the data loaders key(s)
+   * @param valueType - class of the data loaders value(s)
+   * @param name - name of the data loader
+   * @return the breadcrumb
+   */
+  public static @NotNull Breadcrumb graphqlDataLoader(
+      final @NotNull Iterable<?> keys,
+      final @Nullable Class<?> keyType,
+      final @Nullable Class<?> valueType,
+      final @Nullable String name) {
+    final Breadcrumb breadcrumb = new Breadcrumb();
+
+    breadcrumb.setType("graphql");
+    breadcrumb.setCategory("data_loader");
+
+    breadcrumb.setData("keys", keys);
```

Comment:
This also serializes objects used as keys which may lead to very large breadcrumbs.

Reply:
I'd just add the most important info to crumbs since they grow fast, I don't know what `keys` are here so in case they are big and not important to understand which request was performed, I'd leave it away.

Reply:
Probably depends on complexity of the query / response but I'd say it's helpful in figuring out what was being loaded from the data loader. Could also call `toString` on the key and put that in the breadcrumb. By default that'd mean pretty much useless info like `MyObject@123` so not sure that's a good appraoch.

Reply:
`MyObject@123`, nope, that's not useful.
I guess the query would have all the info needed, the data loaders would just load specifically based on the query anyway.
Crumbs don't have the query though, maybe you can make an array of the `keys` but without its values to reduce its size?

Reply:
> I guess the query would have all the info needed, the data loaders would just load specifically based on the query anyway.

I don't think that's always true. There might be changing state in the database resulting in different outcomes depending on when a certain query is sent. So knowing which keys might be the only way to dig into some of the cases. To further illustrate this using our spring dataloader example, we might be asking for a list of tasks with some filter criteria. As data in the database changes we might return different tasks and assignees may also change. So knowing which assignees were queried may only be possible using our breadcrumbs or looking at versioned database entries (if available; that may be rather complex to do).